gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystem.java

```diff
@@ -138,40 +140,31 @@
   public GoogleCloudStorageFileSystem(
       Credential credential, GoogleCloudStorageFileSystemOptions options) throws IOException {
     logger.atFine().log("GoogleCloudStorageFileSystem(options: %s)", options);
-    options.throwIfNotValid();
 
-    checkArgument(credential != null, "credential must not be null");
+    checkNotNull(credential, "credential must not be null");
 
-    this.options = options;
-    this.gcs = new GoogleCloudStorageImpl(options.getCloudStorageOptions(), credential);
+    this.options = checkNotNull(options, "options must not be null");
+    this.options.throwIfNotValid();
+
+    this.gcs = new GoogleCloudStorageImpl(this.options.getCloudStorageOptions(), credential);
 
     if (options.isPerformanceCacheEnabled()) {
       this.gcs =
-          new PerformanceCachingGoogleCloudStorage(this.gcs, options.getPerformanceCacheOptions());
+          new PerformanceCachingGoogleCloudStorage(
+              this.gcs, this.options.getPerformanceCacheOptions());
     }
   }
 
-  /**
-   * Constructs a GoogleCloudStorageFilesystem based on an already-configured underlying
-   * GoogleCloudStorage {@code gcs}.
-   */
-  @VisibleForTesting
-  public GoogleCloudStorageFileSystem(GoogleCloudStorage gcs) {
-    this(
-        gcs,
-        GoogleCloudStorageFileSystemOptions.builder()
-            .setCloudStorageOptions(gcs.getOptions())
-            .build());
-  }
-
   /**
    * Constructs a GoogleCloudStorageFilesystem based on an already-configured underlying
    * GoogleCloudStorage {@code gcs}. Any options pertaining to GCS creation will be ignored.
    */
   @VisibleForTesting
   public GoogleCloudStorageFileSystem(
-      GoogleCloudStorage gcs, GoogleCloudStorageFileSystemOptions options) {
-    this.gcs = gcs;
+      CheckedFunction<GoogleCloudStorageOptions, GoogleCloudStorage, IOException> gcsFn,
+      GoogleCloudStorageFileSystemOptions options)
+      throws IOException {
+    this.gcs = gcsFn.apply(options.getCloudStorageOptions());
```

Comment:
Would it be better to construct GoogleCloudStorageFileSystemOptions and pass in a GCSOption here so that you dont have to do `GoogleCloudStorageFileSystemOptions.builder()
              .setCloudStorageOptions(getInMemoryGoogleCloudStorageOptions())
              .build()` for every caller?

Reply:
Different tests want to create different instances of `CloudStorageInterface`, that's why I do not think that just passing options class will work.